hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -696,3 +696,73 @@ def valid_tuple_axes(ndim, min_size=0, max_size=None):
         lambda x: x if x < ndim else x - 2 * ndim
     )
     return st.lists(axes, min_size, max_size, unique_by=lambda x: x % ndim).map(tuple)
+
+
+def _insert_singleton(mask, shape):
+    # type: (Sequence[bool], Sequence[int]) -> List[int]
+    return [1 if m else s for m, s in zip(mask, reversed(shape))][::-1]
+
+
+def _aligned_shapes(shape, min_dims, max_dims):
+    # type: (Sequence[int], int, int) -> st.SearchStrategy[List[int]]
+    return st.lists(st.booleans(), min_size=min_dims, max_size=max_dims).map(
+        lambda x: _insert_singleton(x, shape)
+    )
+
+
+@st.composite
+def _broadcastable_shapes(draw, shape, min_dims=0, max_dims=3, min_side=0, max_side=5):
```

Comment:
Seeing it laid out as an `@composite` strategy, and the helpers above, this is feeling a bit convoluted to me.  Given the very large impact of array shapes on what data is generated, I think it's worth writing a SearchStrategy class and using the low-level tools:

```python
class BroadcastShapeStrategy(SearchStrategy):

    def __init__(self, shape, min_dims, max_dims, min_side, max_side):
        assert 0 <= min_side <= max_side and 0 <= min_dims <= max_dims <= 32
        SearchStrategy.__init__(self)
        self.shape = shape
        self.side_strat = st.integers(min_side, max_side)
        self.min_dims = min_dims
        self.max_dims = max_dims

    def do_draw(self, data):
        elements = cu.many(
            data,
            min_size=self.min_dims,
            max_size=self.max_dims,
            average_size=self.min_dims,
        )
        result = []
        while elements.more():
            if len(result) < len(shape)
                # Shrinks towards side length 1, rather than shape, if allowed.
                if (min_side > 1) or data.draw(st.booleans()):
                    side = reversed(shape)[len(result)]
                else:
                    side = 1
            else:
                side = data.draw(self.side_strat)
            result.append(side)
        assert min_dims <= len(result) <= max_dims
        assert all(min_side <= s <= max_side for s in result)
        return tuple(reversed(result))
```

Using `cu.many` means we get all our usual properties about collection elements, and can use some more complicated logic internally to choose each.